index.d.ts

```diff
@@ -0,0 +1,27 @@
+interface DeferredPromise<T> {
+	/**
+	Resolves the promise with a value or the result of another promise.
+
+	@param value - The value to resolve the promise with.
+	*/
+	resolve(value: T | PromiseLike<T>): void;
+
+	/**
+	Reject the promise with a provided reason or error.
+
+	@param reason - The reason or error to reject the promise with.
+	*/
+	reject(reason: any): void;
+
+	/**
+	The deferred promise.
+	*/
+	promise: Promise<T>;
```

Comment:
Since `ValueType` is used earlier, we should use it here too for consistency.

```suggestion
	promise: Promise<ValueType>;
```